sharding/sharding.go

```diff
@@ -0,0 +1,311 @@
+			println(message)
+			sr.inconsistencyLog.Printf(message)
+		} else {
+			println(err.Error())
+		}
+	}
+	return resp, err
+}
+
+func newMultiBackendCluster(transp http.RoundTripper,
+	multiResponseHandler transport.MultipleResponsesHandler,
+	clusterConf config.ClusterConfig, name string) cluster {
+	backends := make([]*url.URL, len(clusterConf.Backends))
+
+	for i, backend := range clusterConf.Backends {
+		backends[i] = backend.URL
+	}
+
+	multiTransport := transport.NewMultiTransport(
+		transp,
+		backends,
+		multiResponseHandler)
+
+	return cluster{
+		multiTransport,
+		clusterConf.Weight,
+		clusterConf.Backends,
+		name,
+	}
+}
+
+type ringFactory struct {
+	conf                    config.Config
+	transport               http.RoundTripper
+	multipleResponseHandler transport.MultipleResponsesHandler
+	clusters                map[string]cluster
+}
+
+func (rf ringFactory) initCluster(name string) (cluster, error) {
+	clusterConf, ok := rf.conf.Clusters[name]
+	if !ok {
+		return cluster{}, fmt.Errorf("no cluster %q in configuration", name)
+	}
+	return newMultiBackendCluster(rf.transport, rf.multipleResponseHandler, clusterConf, name), nil
+}
+
+func (rf ringFactory) getCluster(name string) (cluster, error) {
+	s3cluster, ok := rf.clusters[name]
+	if ok {
+		return s3cluster, nil
+	}
+	s3cluster, err := rf.initCluster(name)
+	if err != nil {
+		return s3cluster, err
+	}
+	rf.clusters[name] = s3cluster
+	return s3cluster, nil
+}
+
+func (rf ringFactory) mapShards(weightSum uint, clientCfg config.ClientConfig) (map[string]cluster, error) {
+	shardClusterMap := make(map[string]cluster, clientCfg.ShardsCount)
+	offset := 0
+	for _, name := range clientCfg.Clusters {
+		clientCluster, err := rf.getCluster(name)
+		if err != nil {
+			return shardClusterMap, err
+		}
+		shardsNum := float64(clientCfg.ShardsCount) * float64(clientCluster.weight) / float64(weightSum)
+		for i := offset; i < offset+int(shardsNum); i++ {
+			shardName := fmt.Sprintf("%s-%d", clientCfg.Name, i)
+			shardClusterMap[shardName] = clientCluster
+		}
+		offset += int(shardsNum)
+	}
+	return shardClusterMap, nil
+}
+
+func (rf ringFactory) uniqBackends(clientCfg config.ClientConfig) ([]*url.URL, error) {
+	allBackendsSet := make(map[config.YAMLURL]bool)
+	for _, name := range clientCfg.Clusters {
+		clientCluster, err := rf.getCluster(name)
+		if err != nil {
+			return nil, err
+		}
+		for _, backendURL := range clientCluster.backends {
+			allBackendsSet[backendURL] = true
+		}
+	}
+	var uniqBackendsSlice []*url.URL
+	for url := range allBackendsSet {
+		uniqBackendsSlice = append(uniqBackendsSlice, url.URL)
+	}
+	return uniqBackendsSlice, nil
+}
+
+func (rf ringFactory) clientRing(clientCfg config.ClientConfig) (shardsRing, error) {
+	weightSum := uint(0)
+	clientClusters := make([]cluster, 0, len(clientCfg.Clusters))
+	regressionMap := make(map[string]cluster)
+	var previousCluster cluster
+	for i, name := range clientCfg.Clusters {
```

Comment:
I think this loop does too much. It should be split into two (2) functions each returning one parameter regressionMap  and weightSum